api/src/main/java/org/hawkular/inventory/base/Mutator.java

```diff
@@ -0,0 +1,431 @@
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toSet;
+import static org.hawkular.inventory.api.Action.created;
+import static org.hawkular.inventory.api.Action.deleted;
+import static org.hawkular.inventory.api.Action.updated;
+import static org.hawkular.inventory.api.Relationships.Direction.both;
+import static org.hawkular.inventory.api.Relationships.Direction.outgoing;
+import static org.hawkular.inventory.api.Relationships.WellKnown.contains;
+import static org.hawkular.inventory.api.Relationships.WellKnown.defines;
+import static org.hawkular.inventory.api.filters.Related.by;
+import static org.hawkular.inventory.api.filters.With.id;
+import static org.hawkular.inventory.api.filters.With.type;
+
+/**
+ * @author Lukas Krejci
+ * @since 0.1.0
+ */
+abstract class Mutator<BE, E extends Entity<Blueprint, Update>, Blueprint extends Entity.Blueprint,
+        Update extends AbstractElement.Update> extends Traversal<BE, E> {
+
+    protected Mutator(TraversalContext<BE, E> context) {
+        super(context);
+    }
+
+    /**
+     * Extracts the proposed ID from the blueprint or identifies the ID through some other means.
+     *
+     * @param blueprint the blueprint of the entity to be created
+     * @return the ID to be used for the new entity
+     */
+    protected abstract String getProposedId(Blueprint blueprint);
+
+    /**
+     * A helper method to be used in the implementation of the
+     * {@link org.hawkular.inventory.api.WriteInterface#create(Entity.Blueprint)} method.
+     *
+     * <p>The callers may merely use the returned query and construct a new {@code *Single} instance using it.
+     *
+     * @param blueprint the blueprint of the new entity
+     * @return the query to the newly created entity.
+     */
+    protected final Query doCreate(Blueprint blueprint) {
+        String id = getProposedId(blueprint);
+
+        Query existenceCheck = context.hop().filter().with(id(id)).get();
+
+        Page<BE> results = context.backend.query(existenceCheck, Pager.single());
+
+        if (!results.isEmpty()) {
+            throw new EntityAlreadyExistsException(id, Query.filters(existenceCheck));
+        }
+
+        CanonicalPathAndEntity<BE> parentPath = getCanonicalParentPath();
+
+        NewEntityAndPendingNotifications<E> newEntity;
+        BE containsRel = null;
+        try {
+            BE entityObject = context.backend.persist(id, blueprint);
+
+            if (parentPath.path.isDefined()) {
+                BE parent = parentPath.entity;
+                containsRel = context.backend.relate(parent, entityObject, contains.name(), Collections.emptyMap());
+            }
+
+            newEntity = wireUpNewEntity(entityObject, blueprint, parentPath.path, parentPath.entity);
+
+            context.backend.commit();
+        } catch (Throwable t) {
+            context.backend.rollback();
```

Comment:
Transactions in blueprints are thread-local and start with the first access to the database (read or write).
Which reminds me - we need to do `commit()`s even in the querying calls because otherwise we risk seeing stale data.